File: go/queue_systems.go
package main

import (
	"fmt"
	"strings"
	"sync"
	"time"
)

type PrintJob struct {
	ID       int
	Document string
	Pages    int
	Priority int
	UserID   string
}

type PrintQueue struct {
	jobs []PrintJob
	mu   sync.Mutex
}

func NewPrintQueue() *PrintQueue {
	return &PrintQueue{
		jobs: make([]PrintJob, 0),
	}
}

func (pq *PrintQueue) AddJob(job PrintJob) {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	
	inserted := false
	for i, existingJob := range pq.jobs {
		if job.Priority > existingJob.Priority {
			pq.jobs = append(pq.jobs[:i], append([]PrintJob{job}, pq.jobs[i:]...)...)
			inserted = true
			break
		}
	}
	
	if !inserted {
		pq.jobs = append(pq.jobs, job)
	}
	
	fmt.Printf("Added print job: %s (Priority: %d)\n", job.Document, job.Priority)
}

func (pq *PrintQueue) ProcessNext() *PrintJob {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	
	if len(pq.jobs) == 0 {
		return nil
	}
	
	job := pq.jobs[0]
	pq.jobs = pq.jobs[1:]
	return &job
}

func (pq *PrintQueue) GetStatus() {
	pq.mu.Lock()
	defer pq.mu.Unlock()
	
	fmt.Printf("Print Queue Status - %d jobs pending:\n", len(pq.jobs))
	for i, job := range pq.jobs {
		fmt.Printf("  %d. %s (%d pages, Priority: %d) - User: %s\n", 
			i+1, job.Document, job.Pages, job.Priority, job.UserID)
	}
}

type Task struct {
	ID        int
	Name      string
	Priority  int
	Duration  time.Duration
	CreatedAt time.Time
}

type CPUScheduler struct {
	readyQueue   []Task
	currentTask  *Task
	completedTasks []Task
	mu           sync.RWMutex
	isRunning    bool
}

func NewCPUScheduler() *CPUScheduler {
	return &CPUScheduler{
		readyQueue:     make([]Task, 0),
		completedTasks: make([]Task, 0),
	}
}

func (cs *CPUScheduler) AddTask(task Task) {
	cs.mu.Lock()
	defer cs.mu.Unlock()
	
	task.CreatedAt = time.Now()
	
	inserted := false
	for i, existingTask := range cs.readyQueue {
		if task.Priority > existingTask.Priority {
			cs.readyQueue = append(cs.readyQueue[:i], append([]Task{task}, cs.readyQueue[i:]...)...)
			inserted = true
			break
		}
	}
	
	if !inserted {
		cs.readyQueue = append(cs.readyQueue, task)
	}
	
	fmt.Printf("Task added to scheduler: %s (Priority: %d, Duration: %v)\n", 
		task.Name, task.Priority, task.Duration)
}

func (cs *CPUScheduler) RunScheduler() {
	cs.mu.Lock()
	if cs.isRunning {
		cs.mu.Unlock()
		return
	}
	cs.isRunning = true
	cs.mu.Unlock()

	for {
		cs.mu.Lock()
		if len(cs.readyQueue) == 0 {
			cs.isRunning = false
			cs.mu.Unlock()
			break
		}
		
		task := cs.readyQueue[0]
		cs.readyQueue = cs.readyQueue[1:]
		cs.currentTask = &task
		cs.mu.Unlock()
		
		fmt.Printf("Executing task: %s (Duration: %v)\n", task.Name, task.Duration)
		time.Sleep(task.Duration)
		
		cs.mu.Lock()
		cs.completedTasks = append(cs.completedTasks, task)
		cs.currentTask = nil
		cs.mu.Unlock()
		
		fmt.Printf("Task completed: %s\n", task.Name)
	}
}

func (cs *CPUScheduler) GetStatus() {
	cs.mu.RLock()
	defer cs.mu.RUnlock()
	
	fmt.Printf("CPU Scheduler Status:\n")
	if cs.currentTask != nil {
		fmt.Printf("  Current: %s\n", cs.currentTask.Name)
	} else {
		fmt.Printf("  Current: idle\n")
	}
	
	fmt.Printf("  Ready Queue (%d tasks):\n", len(cs.readyQueue))
	for i, task := range cs.readyQueue {
		fmt.Printf("    %d. %s (Priority: %d)\n", i+1, task.Name, task.Priority)
	}
	
	fmt.Printf("  Completed: %d tasks\n", len(cs.completedTasks))
}

type WebPage struct {
	URL      string
	Content  string
	Links    []string
	Depth    int
	Visited  bool
}

type WebCrawler struct {
	queue       []WebPage
	visited     map[string]bool
	maxDepth    int
	mu          sync.RWMutex
	crawledData []WebPage
}

func NewWebCrawler(maxDepth int) *WebCrawler {
	return &WebCrawler{
		queue:       make([]WebPage, 0),
		visited:     make(map[string]bool),
		maxDepth:    maxDepth,
		crawledData: make([]WebPage, 0),
	}
}

func (wc *WebCrawler) AddURL(url string, depth int) {
	wc.mu.Lock()
	defer wc.mu.Unlock()
	
	if wc.visited[url] || depth > wc.maxDepth {
		return
	}
	
	page := WebPage{
		URL:     url,
		Depth:   depth,
		Visited: false,
	}
	
	wc.queue = append(wc.queue, page)
	wc.visited[url] = true
	fmt.Printf("Added to crawl queue: %s (depth: %d)\n", url, depth)
}

func (wc *WebCrawler) simulateFetchPage(url string) WebPage {
	time.Sleep(100 * time.Millisecond)
	
	mockPages := map[string]WebPage{
		"https://example.com": {
			URL:     "https://example.com",
			Content: "Welcome to Example.com - Home page content",
			Links:   []string{"https://example.com/about", "https://example.com/products"},
		},
		"https://example.com/about": {
			URL:     "https://example.com/about",
			Content: "About us page content",
			Links:   []string{"https://example.com/contact", "https://example.com/team"},
		},
		"https://example.com/products": {
			URL:     "https://example.com/products",
			Content: "Our products page content",
			Links:   []string{"https://example.com/product/1", "https://example.com/product/2"},
		},
		"https://example.com/contact": {
			URL:     "https://example.com/contact",
			Content: "Contact us page content",
			Links:   []string{},
		},
		"https://example.com/team": {
			URL:     "https://example.com/team",
			Content: "Meet our team page content",
			Links:   []string{},
		},
		"https://example.com/product/1": {
			URL:     "https://example.com/product/1",
			Content: "Product 1 details",
			Links:   []string{},
		},
		"https://example.com/product/2": {
			URL:     "https://example.com/product/2",
			Content: "Product 2 details",
			Links:   []string{},
		},
	}
	
	if page, exists := mockPages[url]; exists {
		return page
	}
	
	return WebPage{
		URL:     url,
		Content: "Page not found",
		Links:   []string{},
	}
}

func (wc *WebCrawler) Crawl() {
	for {
		wc.mu.Lock()
		if len(wc.queue) == 0 {
			wc.mu.Unlock()
			break
		}
		
		currentPage := wc.queue[0]
		wc.queue = wc.queue[1:]
		wc.mu.Unlock()
		
		fmt.Printf("Crawling: %s (depth: %d)\n", currentPage.URL, currentPage.Depth)
		
		fetchedPage := wc.simulateFetchPage(currentPage.URL)
		fetchedPage.Depth = currentPage.Depth
		fetchedPage.Visited = true
		
		wc.mu.Lock()
		wc.crawledData = append(wc.crawledData, fetchedPage)
		wc.mu.Unlock()
		
		for _, link := range fetchedPage.Links {
			wc.AddURL(link, currentPage.Depth+1)
		}
		
		fmt.Printf("  Found %d links on %s\n", len(fetchedPage.Links), currentPage.URL)
	}
}

func (wc *WebCrawler) GetResults() {
	wc.mu.RLock()
	defer wc.mu.RUnlock()
	
	fmt.Printf("\nCrawl Results - %d pages crawled:\n", len(wc.crawledData))
	for _, page := range wc.crawledData {
		contentPreview := page.Content
		if len(contentPreview) > 50 {
			contentPreview = contentPreview[:50] + "..."
		}
		fmt.Printf("  [Depth %d] %s\n", page.Depth, page.URL)
		fmt.Printf("    Content: %s\n", contentPreview)
		fmt.Printf("    Links found: %d\n", len(page.Links))
	}
}

func demonstrateQueues() {
	fmt.Println("=== Print Spooling Queue Example ===")
	printQueue := NewPrintQueue()
	
	jobs := []PrintJob{
		{1, "Resume.pdf", 2, 1, "alice"},
		{2, "Report.docx", 10, 3, "bob"},
		{3, "Invoice.pdf", 1, 2, "charlie"},
		{4, "Manual.pdf", 50, 1, "david"},
		{5, "Presentation.pptx", 15, 3, "eve"},
	}
	
	for _, job := range jobs {
		printQueue.AddJob(job)
	}
	
	printQueue.GetStatus()
	
	fmt.Println("\nProcessing print jobs:")
	for {
		job := printQueue.ProcessNext()
		if job == nil {
			break
		}
		fmt.Printf("Printing: %s (%d pages) for %s\n", job.Document, job.Pages, job.UserID)
		time.Sleep(500 * time.Millisecond)
	}

	fmt.Println("\n=== CPU Task Scheduling Example ===")
	scheduler := NewCPUScheduler()
	
	tasks := []Task{
		{1, "System Update", 2, 1 * time.Second, time.Time{}},
		{2, "File Backup", 1, 2 * time.Second, time.Time{}},
		{3, "Virus Scan", 3, 1500 * time.Millisecond, time.Time{}},
		{4, "Email Sync", 2, 800 * time.Millisecond, time.Time{}},
		{5, "Database Cleanup", 1, 1200 * time.Millisecond, time.Time{}},
	}
	
	for _, task := range tasks {
		scheduler.AddTask(task)
	}
	
	scheduler.GetStatus()
	
	fmt.Println("\nStarting task execution:")
	scheduler.RunScheduler()
	
	scheduler.GetStatus()

	fmt.Println("\n=== Web Crawler BFS Example ===")
	crawler := NewWebCrawler(2)
	
	crawler.AddURL("https://example.com", 0)
	
	fmt.Println("Starting web crawl...")
	crawler.Crawl()
	
	crawler.GetResults()
}

func main() {
	demonstrateQueues()
}

File: python/queue_systems.py
import asyncio
import time
from collections import deque
from datetime import datetime
from typing import List, Dict, Optional, Set
import heapq

class PrintJob:
    def __init__(self, job_id: int, document: str, pages: int, priority: int, user_id: str):
        self.id = job_id
        self.document = document
        self.pages = pages
        self.priority = priority
        self.user_id = user_id
        self.timestamp = datetime.now()
    
    def __lt__(self, other):
        # Higher priority jobs come first (reverse order for heapq)
        return self.priority > other.priority

class PrintQueue:
    def __init__(self):
        self.jobs = []  # Using heapq for priority queue
        self.job_counter = 0
    
    def add_job(self, job: PrintJob) -> None:
        # Use counter to maintain insertion order for equal priorities
        heapq.heappush(self.jobs, (job, self.job_counter))
        self.job_counter += 1
        print(f"Added print job: {job.document} (Priority: {job.priority})")
    
    def process_next(self) -> Optional[PrintJob]:
        if not self.jobs:
            return None
        
        job, _ = heapq.heappop(self.jobs)
        return job
    
    def get_status(self) -> None:
        print(f"Print Queue Status - {len(self.jobs)} jobs pending:")
        # Sort jobs by priority for display (without modifying the heap)
        sorted_jobs = sorted([job for job, _ in self.jobs], key=lambda x: (-x.priority, x.timestamp))
        
        for i, job in enumerate(sorted_jobs, 1):
            print(f"  {i}. {job.document} ({job.pages} pages, Priority: {job.priority}) - User: {job.user_id}")
    
    def cancel_job(self, job_id: int) -> bool:
        """Cancel a job by ID"""
        for i, (job, counter) in enumerate(self.jobs):
            if job.id == job_id:
                print(f"Canceled job: {job.document}")
                self.jobs[i] = (self.jobs[-1][0], self.jobs[-1][1])  # Replace with last element
                self.jobs.pop()
                heapq.heapify(self.jobs)  # Rebuild heap
                return True
        return False
    
    def get_estimated_wait_time(self, target_job_id: int) -> int:
        """Estimate wait time for a specific job in minutes"""
        wait_time = 0
        for job, _ in sorted(self.jobs, key=lambda x: (-x[0].priority, x[0].timestamp)):
            if job.id == target_job_id:
                break
            wait_time += job.pages * 0.5  # Assume 30 seconds per page
        return int(wait_time)

class Task:
    def __init__(self, task_id: int, name: str, priority: int, duration: float, dependencies: List[int] = None):
        self.id = task_id
        self.name = name
        self.priority = priority
        self.duration = duration  # in seconds
        self.dependencies = dependencies or []
        self.created_at = datetime.now()
    
    def __lt__(self, other):
        return self.priority > other.priority

class CPUScheduler:
    def __init__(self):
        self.ready_queue = []  # Priority queue
        self.current_task: Optional[Task] = None
        self.completed_tasks: List[Task] = []
        self.waiting_tasks: List[Task] = []  # Tasks waiting for dependencies
        self.is_running = False
        self.task_counter = 0
    
    def add_task(self, task: Task) -> None:
        if self._dependencies_met(task):
            heapq.heappush(self.ready_queue, (task, self.task_counter))
        else:
            self.waiting_tasks.append(task)
        
        self.task_counter += 1
        deps_str = f" (depends on: {task.dependencies})" if task.dependencies else ""
        print(f"Task added to scheduler: {task.name} (Priority: {task.priority}, Duration: {task.duration}s){deps_str}")
    
    def _dependencies_met(self, task: Task) -> bool:
        if not task.dependencies:
            return True
        
        completed_ids = {t.id for t in self.completed_tasks}
        return all(dep_id in completed_ids for dep_id in task.dependencies)
    
    def _check_waiting_tasks(self) -> None:
        """Move tasks from waiting to ready queue if dependencies are met"""
        ready_tasks = []
        still_waiting = []
        
        for task in self.waiting_tasks:
            if self._dependencies_met(task):
                ready_tasks.append(task)
            else:
                still_waiting.append(task)
        
        for task in ready_tasks:
            heapq.heappush(self.ready_queue, (task, self.task_counter))
            self.task_counter += 1
            print(f"Task {task.name} moved to ready queue (dependencies met)")
        
        self.waiting_tasks = still_waiting
    
    async def run_scheduler(self) -> None:
        if self.is_running:
            print("Scheduler is already running")
            return
        
        self.is_running = True
        print("Starting CPU scheduler...")
        
        while self.ready_queue or self.waiting_tasks:
            self._check_waiting_tasks()
            
            if not self.ready_queue:
                if self.waiting_tasks:
                    print("All remaining tasks are waiting for dependencies...")
                    break
                else:
                    break
            
            task, _ = heapq.heappop(self.ready_queue)
            self.current_task = task
            
            print(f"Executing task: {task.name} (Duration: {task.duration}s)")
            start_time = time.time()
            
            await asyncio.sleep(task.duration)
            
            execution_time = time.time() - start_time
            print(f"Task completed: {task.name} (actual time: {execution_time:.2f}s)")
            
            self.completed_tasks.append(task)
            self.current_task = None
        
        self.is_running = False
        if self.waiting_tasks:
            print(f"Warning: {len(self.waiting_tasks)} tasks remain with unmet dependencies")
        print("All executable tasks completed")
    
    def get_status(self) -> None:
        print("CPU Scheduler Status:")
        print(f"  Current: {self.current_task.name if self.current_task else 'idle'}")
        
        ready_tasks = sorted([task for task, _ in self.ready_queue], key=lambda x: (-x.priority, x.created_at))
        print(f"  Ready Queue ({len(ready_tasks)} tasks):")
        for i, task in enumerate(ready_tasks, 1):
            print(f"    {i}. {task.name} (Priority: {task.priority})")
        
        if self.waiting_tasks:
            print(f"  Waiting for dependencies ({len(self.waiting_tasks)} tasks):")
            for i, task in enumerate(self.waiting_tasks, 1):
                print(f"    {i}. {task.name} (waiting for: {task.dependencies})")
        
        print(f"  Completed: {len(self.completed_tasks)} tasks")
    
    def get_average_wait_time(self) -> float:
        if not self.completed_tasks:
            return 0.0
        
        total_wait_time = sum(
            (datetime.now() - task.created_at).total_seconds() 
            for task in self.completed_tasks
        )
        return total_wait_time / len(self.completed_tasks)

class WebPage:
    def __init__(self, url: str, content: str = "", links: List[str] = None, depth: int = 0):
        self.url = url
        self.content = content
        self.links = links or []
        self.depth = depth
        self.timestamp = datetime.now()

class WebCrawler:
    def __init__(self, max_depth: int = 3, max_pages: int = 50, delay: float = 0.1):
        self.queue = deque()  # BFS queue
        self.visited: Set[str] = set()
        self.crawled_data: List[WebPage] = []
        self.max_depth = max_depth
        self.max_pages = max_pages
        self.delay = delay  # Delay between requests
        self.robots_txt_rules: Dict[str, List[str]] = {}
    
    def add_url(self, url: str, depth: int = 0) -> None:
        if (url in self.visited or 
            depth > self.max_depth or 
            len(self.crawled_data) >= self.max_pages):
            return
        
        self.queue.append((url, depth))
        self.visited.add(url)
        print(f"Added to crawl queue: {url} (depth: {depth})")
    
    async def _simulate_fetch_page(self, url: str) -> WebPage:
        """Simulate fetching a web page with realistic data"""
        await asyncio.sleep(self.delay)  # Simulate network delay
        
        # Mock page database
        mock_pages = {
            'https://example.com': WebPage(
                url='https://example.com',
                content='Welcome to Example.com - Your trusted source for innovative solutions and cutting-edge technology',
                links=['https://example.com/about', 'https://example.com/products', 'https://example.com/blog', 'https://example.com/contact']
            ),
            'https://example.com/about': WebPage(
                url='https://example.com/about',
                content='About Example.com - Founded in 2010, we are a leading technology company specializing in web solutions',
                links=['https://example.com/team', 'https://example.com/careers', 'https://example.com/history']
            ),
            'https://example.com/products': WebPage(
                url='https://example.com/products',
                content='Our Products - Discover our comprehensive suite of software solutions designed for modern businesses',
                links=['https://example.com/product/analytics', 'https://example.com/product/crm', 'https://example.com/product/api']
            ),
            'https://example.com/blog': WebPage(
                url='https://example.com/blog',
                content='Company Blog - Stay updated with the latest industry trends, insights, and company news',
                links=['https://example.com/blog/tech-trends-2024', 'https://example.com/blog/ai-revolution', 'https://example.com/blog/web-security']
            ),
            'https://example.com/contact': WebPage(
                url='https://example.com/contact',
                content='Contact Us - Get in touch with our expert team for support, sales, or partnership opportunities',
                links=['https://example.com/support', 'https://example.com/sales']
            ),
            'https://example.com/team': WebPage(
                url='https://example.com/team',
                content='Our Team - Meet the talented individuals behind our innovative products and services',
                links=[]
            ),
            'https://example.com/product/analytics': WebPage(
                url='https://example.com/product/analytics',
                content='Analytics Platform - Advanced data analytics and visualization tools for business intelligence',
                links=['https://example.com/product/analytics/features', 'https://example.com/product/analytics/pricing']
            ),
            'https://example.com/blog/tech-trends-2024': WebPage(
                url='https://example.com/blog/tech-trends-2024',
                content='Technology Trends 2024 - Exploring the most significant technological developments shaping the future',
                links=[]
            )
        }
        
        if url in mock_pages:
            page = mock_pages[url]
            # Filter links to stay within domain and respect depth
            valid_links = [link for link in page.links if link.startswith('https://example.com')]
            page.links = valid_links
            return page
        
        return WebPage(
            url=url,
            content='Page not found or unable to fetch content',
            links=[]
        )
    
    def _is_allowed_by_robots(self, url: str) -> bool:
        """Simulate robots.txt checking"""
        # In a real crawler, this would parse robots.txt
        disallowed_paths = ['/admin', '/private', '/internal']
        return not any(path in url for path in disallowed_paths)
    
    async def crawl(self, concurrent_requests: int = 3) -> None:
        print("Starting web crawl...")
        
        semaphore = asyncio.Semaphore(concurrent_requests)
        
        while self.queue and len(self.crawled_data) < self.max_pages:
            # Process multiple URLs concurrently
            batch_size = min(concurrent_requests, len(self.queue))
            batch = []
            
            for _ in range(batch_size):
                if self.queue:
                    batch.append(self.queue.popleft())
            
            tasks = [self._crawl_single_page(url, depth, semaphore) for url, depth in batch]
            await asyncio.gather(*tasks, return_exceptions=True)
        
        print(f"Web crawl completed - {len(self.crawled_data)} pages crawled")
    
    async def _crawl_single_page(self, url: str, depth: int, semaphore: asyncio.Semaphore) -> None:
        async with semaphore:
            if not self._is_allowed_by_robots(url):
                print(f"Skipping {url} (blocked by robots.txt)")
                return
            
            print(f"Crawling: {url} (depth: {depth})")
            
            try:
                page = await self._simulate_fetch_page(url)
                page.depth = depth
                
                self.crawled_data.append(page)
                
                # Add discovered links to queue
                for link in page.links:
                    if link not in self.visited:
                        self.add_url(link, depth + 1)
                
                print(f"  Found {len(page.links)} links on {url}")
                
            except Exception as e:
                print(f"Error crawling {url}: {e}")
    
    def get_results(self) -> List[WebPage]:
        print(f"\nCrawl Results - {len(self.crawled_data)} pages crawled:")
        
        for page in self.crawled_data:
            content_preview = (page.content[:80] + '...') if len(page.content) > 80 else page.content
            print(f"  [Depth {page.depth}] {page.url}")
            print(f"    Content: {content_preview}")
            print(f"    Links found: {len(page.links)}")
            print(f"    Crawled at: {page.timestamp.strftime('%H:%M:%S')}")
        
        return self.crawled_data
    
    def get_statistics(self) -> Dict[str, float]:
        if not self.crawled_data:
            return {"total_pages": 0, "avg_depth": 0, "total_links": 0, "pages_per_depth": {}}
        
        total_pages = len(self.crawled_data)
        avg_depth = sum(page.depth for page in self.crawled_data) / total_pages
        total_links = sum(len(page.links) for page in self.crawled_data)
        
        # Pages per depth level
        depth_counts = {}
        for page in self.crawled_data:
            depth_counts[page.depth] = depth_counts.get(page.depth, 0) + 1
        
        return {
            "total_pages": total_pages,
            "avg_depth": round(avg_depth, 2),
            "total_links": total_links,
            "pages_per_depth": depth_counts,
            "avg_links_per_page": round(total_links / total_pages, 2)
        }
    
    def export_sitemap(self) -> List[str]:
        """Export discovered URLs as a sitemap"""
        return sorted([page.url for page in self.crawled_data])

async def demonstrate_queues():
    print("=== Print Spooling Queue Example ===")
    print_queue = PrintQueue()
    
    jobs = [
        PrintJob(1, "Resume.pdf", 2, 1, "alice"),
        PrintJob(2, "Report.docx", 10, 3, "bob"),
        PrintJob(3, "Invoice.pdf", 1, 2, "charlie"),
        PrintJob(4, "Manual.pdf", 50, 1, "david"),
        PrintJob(5, "Presentation.pptx", 15, 3, "eve"),
        PrintJob(6, "Urgent_Contract.pdf", 5, 4, "frank")  # Highest priority
    ]
    
    for job in jobs:
        print_queue.add_job(job)
    
    print_queue.get_status()
    
    print(f"\nEstimated wait time for job 4: {print_queue.get_estimated_wait_time(4)} minutes")
    
    print("\nProcessing print jobs:")
    while True:
        job = print_queue.process_next()
        if not job:
            break
        print(f"Printing: {job.document} ({job.pages} pages) for {job.user_id}")
        await asyncio.sleep(0.5)  # Simulate printing time

    print("\n=== CPU Task Scheduling Example ===")
    scheduler = CPUScheduler()
    
    tasks = [
        Task(1, "System Initialization", 3, 0.8),
        Task(2, "Load Configuration", 2, 0.5, dependencies=[1]),
        Task(3, "Start Database", 3, 1.2, dependencies=[1]),
        Task(4, "Start Web Server", 2, 0.7, dependencies=[2, 3]),
        Task(5, "Background Cleanup", 1, 1.5),
        Task(6, "Health Check", 2, 0.3, dependencies=[4]),
        Task(7, "Send Notifications", 1, 0.4, dependencies=[4])
    ]
    
    for task in tasks:
        scheduler.add_task(task)
    
    scheduler.get_status()
    
    print("\nStarting task execution:")
    await scheduler.run_scheduler()
    
    scheduler.get_status()
    print(f"Average wait time: {scheduler.get_average_wait_time():.2f} seconds")

    print("\n=== Web Crawler BFS Example ===")
    crawler = WebCrawler(max_depth=2, max_pages=15, delay=0.1)
    
    crawler.add_url('https://example.com', 0)
    await crawler.crawl(concurrent_requests=2)
    
    results = crawler.get_results()
    stats = crawler.get_statistics()
    
    print("\nCrawl Statistics:")
    for key, value in stats.items():
        if key != "pages_per_depth":
            print(f"  {key.replace('_', ' ').title()}: {value}")
        else:
            print(f"  Pages per depth: {dict(value)}")
    
    print(f"\nSitemap ({len(crawler.export_sitemap())} URLs):")
    for url in crawler.export_sitemap()[:10]:  # Show first 10
        print(f"  {url}")

if __name__ == "__main__":
    asyncio.run(demonstrate_queues())